crates/holochain/src/core/workflow/sys_validation_workflow/validation_deps.rs

```diff
@@ -0,0 +1,203 @@
+        self.states
+            .get(hash)
+            .map(|state| state.dependency.is_some())
+            .unwrap_or(false)
+    }
+
+    /// Get the state of a given dependency. This should always return a value because we should know about the depdendency
+    /// by examining the ops that are being validated. However, the dependency may not be found on the DHT yet.
+    pub fn get(&mut self, hash: &ActionHash) -> Option<&mut ValidationDependencyState> {
+        match self.states.get_mut(hash) {
+            Some(dep) => Some(dep),
+            None => {
+                tracing::warn!(hash = ?hash, "Have not attempted to fetch requested dependency, this is a bug");
+                None
+            }
+        }
+    }
+
+    /// Get the hashes of all dependencies that are currently missing from the DHT.
+    pub fn get_missing_hashes(&self) -> Vec<ActionHash> {
+        self.states
+            .iter()
+            .filter_map(|(hash, state)| {
+                if state.dependency.is_none() {
+                    Some(hash.clone())
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    /// Get the hashes of all dependencies that have been fetched from the network.
+    /// We need to let the incoming dht ops workflow know about these so that it can ingest them and get them validated.
+    pub fn get_network_fetched_hashes(&self) -> Vec<ActionHash> {
+        self.states
+            .iter()
+            .filter_map(|(hash, state)| match state {
+                ValidationDependencyState {
+                    dependency:
+                        Some(ValidationDependency {
+                            fetched_from: CascadeSource::Network,
+                            ..
+                        }),
+                    ..
+                } => Some(hash.clone()),
+                _ => None,
+            })
+            .collect()
+    }
+
+    /// Insert a record which was found after this set of dependencies was created.
+    pub fn insert(&mut self, action: SignedActionHashed, source: CascadeSource) -> bool {
+        let hash = action.as_hash();
+
+        // Note that `has` is checking that the dependency is actually set, not just that we have the key!
+        if self.has(hash) {
+            tracing::warn!(hash = ?hash, "Attempted to insert a dependency that was already present, this is not expected");
+            return false;
+        }
+
+        self.retained_deps.insert(hash.clone());
+
+        if let Some(s) = self.states.get_mut(hash) {
+            s.set_action(action);
+            s.set_source(source);
+            return true;
+        }
+
+        false
+    }
+
+    /// Forget which dependencies have been accessed since this method was last called.
+    /// This is intended to be used with [`purge_held_deps`] to remove any dependencies that are no longer needed.
+    pub fn clear_retained_deps(&mut self) {
```

Comment:
I don't understand the difference yet between this and `purge`. If I'm supposed to clear first and then purge only retains what's in retained_deps which is empty, states must be empty too.

Reply:
The object is being retained between runs and I want to hold onto as much as possible to avoid new cascade calls. So I'm saying at the start of a workflow run, let's forget which things were needed. Run through all the ops that need validation and check their dependencies. Anything that we already have in memory we filter out. The act of checking if we already hold something marks it as used again. Anything that we no longer need can be dropped.

This is partly because we aren't loading deps multiple times if they're needed my multiple ops. So for example if a Create action is both the previous and the thing that's being deleted, then we'll only load it once. So the easiest way to tell if it's not used is to check whether it got accessed as part of checking all our ops which are currently outstanding for validation.